File: sources/net.sf.j2s.core/test/src/com/falstad/circuit/CounterElm.js
(function(){var P$=Clazz.newPackage("com.falstad.circuit"),I$=[];
var C$=Clazz.newClass(P$, "CounterElm", null, 'com.falstad.circuit.ChipElm');

C$.$clinit$ = function() {Clazz.load(C$, 1);
}

Clazz.newMeth(C$, '$init0$', function () {
var c;if((c = C$.superclazz) && (c = c.$init0$))c.apply(this);
this.FLAG_ENABLE = 0;
this.invertreset = false;
}, 1);

Clazz.newMeth(C$, '$init$', function () {
this.FLAG_ENABLE = 2;
this.invertreset = false;
}, 1);

Clazz.newMeth(C$, 'c$$I$I', function (xx, yy) {
C$.superclazz.c$$I$I.apply(this, [xx, yy]);
C$.$init$.apply(this);
}, 1);

Clazz.newMeth(C$, 'c$$I$I$I$I$I$java_util_StringTokenizer', function (xa, ya, xb, yb, f, st) {
C$.superclazz.c$$I$I$I$I$I$java_util_StringTokenizer.apply(this, [xa, ya, xb, yb, f, st]);
C$.$init$.apply(this);
if (st.hasMoreTokens()) this.invertreset =  Boolean.from(st.nextToken()).booleanValue();
 else this.invertreset = true;
this.pins[1].bubble = this.invertreset;
}, 1);

Clazz.newMeth(C$, 'dump', function () {
return C$.superclazz.prototype.dump.apply(this, []) + " " + this.invertreset ;
});

Clazz.newMeth(C$, 'needsBits', function () {
return true;
});

Clazz.newMeth(C$, 'getChipName', function () {
return "Counter";
});

Clazz.newMeth(C$, 'setupPins', function () {
this.sizeX = 2;
this.sizeY = this.bits > 2 ? this.bits : 2;
this.pins = Clazz.array((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))), [this.getPostCount()]);
this.pins[0] = Clazz.new_((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))).c$$I$I$S, [this, null, 0, 2, ""]);
this.pins[0].clock = true;
this.pins[1] = Clazz.new_((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))).c$$I$I$S, [this, null, this.sizeY - 1, 2, "R"]);
this.pins[1].bubble = this.invertreset;
var i;
for (i = 0; i != this.bits; i++) {
var ii = i + 2;
this.pins[ii] = Clazz.new_((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))).c$$I$I$S, [this, null, i, 3, "Q" + (this.bits - i - 1 )]);
this.pins[ii].output = this.pins[ii].state = true;
}
if (this.hasEnable()) this.pins[this.bits + 2] = Clazz.new_((I$[0]||(I$[0]=Clazz.load(Clazz.load('com.falstad.circuit.ChipElm').Pin))).c$$I$I$S, [this, null, this.sizeY - 2, 2, "En"]);
this.allocNodes();
});

Clazz.newMeth(C$, 'getPostCount', function () {
if (this.hasEnable()) return this.bits + 3;
return this.bits + 2;
});

Clazz.newMeth(C$, 'getEditInfo$I', function (n) {
if (n == 0) {
var ei = Clazz.new_((I$[1]||(I$[1]=Clazz.load('com.falstad.circuit.EditInfo'))).c$$S$D$D$D,["", 0, -1, -1]);
ei.checkbox = Clazz.new_((I$[2]||(I$[2]=Clazz.load('a2s.Checkbox'))).c$$S$Z,["Flip X", (this.flags & 1024) != 0]);
return ei;
}if (n == 1) {
var ei = Clazz.new_((I$[1]||(I$[1]=Clazz.load('com.falstad.circuit.EditInfo'))).c$$S$D$D$D,["", 0, -1, -1]);
ei.checkbox = Clazz.new_((I$[2]||(I$[2]=Clazz.load('a2s.Checkbox'))).c$$S$Z,["Flip Y", (this.flags & 2048) != 0]);
return ei;
}if (n == 2) {
var ei = Clazz.new_((I$[1]||(I$[1]=Clazz.load('com.falstad.circuit.EditInfo'))).c$$S$D$D$D,["", 0, -1, -1]);
ei.checkbox = Clazz.new_((I$[2]||(I$[2]=Clazz.load('a2s.Checkbox'))).c$$S$Z,["Invert reset pin", this.invertreset]);
return ei;
}return null;
});

Clazz.newMeth(C$, 'setEditValue$I$com_falstad_circuit_EditInfo', function (n, ei) {
if (n == 0) {
if (ei.checkbox.getState()) this.flags = this.flags|(1024);
 else this.flags = this.flags&(-1025);
this.setPoints();
}if (n == 1) {
if (ei.checkbox.getState()) this.flags = this.flags|(2048);
 else this.flags = this.flags&(-2049);
this.setPoints();
}if (n == 2) {
if (ei.checkbox.getState()) {
this.invertreset = true;
this.pins[1].bubble = true;
} else {
this.invertreset = false;
this.pins[1].bubble = false;
}this.setPoints();
}});

Clazz.newMeth(C$, 'hasEnable', function () {
return (this.flags & 2) != 0;
});

Clazz.newMeth(C$, 'getVoltageSourceCount', function () {
return this.bits;
});

Clazz.newMeth(C$, 'execute', function () {
var en = true;
if (this.hasEnable()) en = this.pins[this.bits + 2].value;
if (this.pins[0].value && !this.lastClock && en  ) {
var i;
for (i = this.bits - 1; i >= 0; i--) {
var ii = i + 2;
if (!this.pins[ii].value) {
this.pins[ii].value = true;
break;
}this.pins[ii].value = false;
}
}if (!this.pins[1].value == this.invertreset ) {
var i;
for (i = 0; i != this.bits; i++) this.pins[i + 2].value = false;

}this.lastClock = this.pins[0].value;
});

Clazz.newMeth(C$, 'getDumpType', function () {
return 164;
});

Clazz.newMeth(C$);
})();
//Created 2017-12-17 19:28:18

File: sources/net.sf.j2s.java.core/src2/com/falstad/circuit/CounterElm.java
package com.falstad.circuit;


import java.util.StringTokenizer;

import java.awt.Checkbox;

class CounterElm extends ChipElm {
	final int FLAG_ENABLE = 2;
	boolean invertreset = false;

	public CounterElm(int xx, int yy) {
		super(xx, yy);
	}

	public CounterElm(int xa, int ya, int xb, int yb, int f, StringTokenizer st) {
		super(xa, ya, xb, yb, f, st);
		if (st.hasMoreTokens())
			invertreset = new Boolean(st.nextToken()).booleanValue();
		else
			invertreset = true;
		pins[1].bubble = invertreset;
	}

	String dump() {
		return super.dump() + " " + invertreset;
	}

	boolean needsBits() {
		return true;
	}

	String getChipName() {
		return "Counter";
	}

	void setupPins() {
		sizeX = 2;
		sizeY = bits > 2 ? bits : 2;
		pins = new Pin[getPostCount()];
		pins[0] = new Pin(0, SIDE_W, "");
		pins[0].clock = true;
		pins[1] = new Pin(sizeY - 1, SIDE_W, "R");
		pins[1].bubble = invertreset;
		int i;
		for (i = 0; i != bits; i++) {
			int ii = i + 2;
			pins[ii] = new Pin(i, SIDE_E, "Q" + (bits - i - 1));
			pins[ii].output = pins[ii].state = true;
		}
		if (hasEnable())
			pins[bits + 2] = new Pin(sizeY - 2, SIDE_W, "En");
		allocNodes();
	}

	int getPostCount() {
		if (hasEnable())
			return bits + 3;
		return bits + 2;
	}

	public EditInfo getEditInfo(int n) {
		if (n == 0) {
			EditInfo ei = new EditInfo("", 0, -1, -1);
			ei.checkbox = new Checkbox("Flip X", (flags & FLAG_FLIP_X) != 0);
			return ei;
		}
		if (n == 1) {
			EditInfo ei = new EditInfo("", 0, -1, -1);
			ei.checkbox = new Checkbox("Flip Y", (flags & FLAG_FLIP_Y) != 0);
			return ei;
		}
		if (n == 2) {
			EditInfo ei = new EditInfo("", 0, -1, -1);
			ei.checkbox = new Checkbox("Invert reset pin", invertreset);
			return ei;
		}
		return null;
	}

	public void setEditValue(int n, EditInfo ei) {
		if (n == 0) {
			if (ei.checkbox.getState())
				flags |= FLAG_FLIP_X;
			else
				flags &= ~FLAG_FLIP_X;
			setPoints();
		}
		if (n == 1) {
			if (ei.checkbox.getState())
				flags |= FLAG_FLIP_Y;
			else
				flags &= ~FLAG_FLIP_Y;
			setPoints();
		}
		if (n == 2) {
			if (ei.checkbox.getState()) {
				invertreset = true;
				pins[1].bubble = true;
			} else {
				invertreset = false;
				pins[1].bubble = false;
			}
			setPoints();
		}
	}

	boolean hasEnable() {
		return (flags & FLAG_ENABLE) != 0;
	}

	int getVoltageSourceCount() {
		return bits;
	}

	void execute() {
		boolean en = true;
		if (hasEnable())
			en = pins[bits + 2].value;
		if (pins[0].value && !lastClock && en) {
			int i;
			for (i = bits - 1; i >= 0; i--) {
				int ii = i + 2;
				if (!pins[ii].value) {
					pins[ii].value = true;
					break;
				}
				pins[ii].value = false;
			}
		}
		if (!pins[1].value == invertreset) {
			int i;
			for (i = 0; i != bits; i++)
				pins[i + 2].value = false;
		}
		lastClock = pins[0].value;
	}

	int getDumpType() {	return 164;
	}
}
